lib/Optimizer/Transforms/ClassicalOptimization.cpp

```diff
@@ -0,0 +1,168 @@
+
+  void runOnOperation() override {
+    auto *ctx = &getContext();
+    auto *op = getOperation();
+    DominanceInfo domInfo(op);
+    auto func = dyn_cast<func::FuncOp>(op);
+    auto numLoops = countLoopOps(op);
+    unsigned progress = 0;
+
+    RewritePatternSet patterns(ctx);
+    for (auto *dialect : ctx->getLoadedDialects())
+      dialect->getCanonicalizationPatterns(patterns);
+    for (RegisteredOperationName op : ctx->getRegisteredOperations())
+      op.getCanonicalizationPatterns(patterns, ctx);
+
+    // Add patterns that help const prop loop boundaries computed
+    // in conditional statements, other loops, or arrays.
+    patterns.insert<RewriteIf>(ctx, /*rewriteOnlyIfConst=*/true);
+    patterns.insert<LoopPat>(ctx, allowClosedInterval, allowBreak);
+    patterns.insert<AllocaPattern>(
+        ctx, domInfo, func == nullptr ? "unknown" : func.getName());
+    if (numLoops)
+      patterns.insert<UnrollCountedLoop>(ctx, threshold,
+                                         /*signalFailure=*/false, allowBreak,
+                                         progress);
+
+    FrozenRewritePatternSet frozen(std::move(patterns));
+    // Iterate over the loops until a fixed-point is reached. Some loops can
+    // only be unrolled if other loops are unrolled first and the constants
+    // iteratively propagated.
+    do {
+      // Remove overridden writes.
+      auto analysis = SimplifyWritesAnalysis(domInfo, op);
+      analysis.removeOverriddenStores();
+      // Clean up dead code.
+      {
+        auto builder = OpBuilder(op);
+        IRRewriter rewriter(builder);
+        [[maybe_unused]] auto unused =
+            simplifyRegions(rewriter, op->getRegions());
+      }
+      progress = 0;
+      (void)applyPatternsAndFoldGreedily(op, frozen);
+    } while (progress);
+
+    if (signalFailure) {
+      numLoops = countLoopOps(op);
+      if (numLoops) {
+        op->emitOpError("did not unroll loops");
+        signalPassFailure();
+      }
+    }
+  }
+
+  static unsigned countLoopOps(Operation *op) {
+    unsigned result = 0;
+    op->walk([&](cudaq::cc::LoopOp loop) { result++; });
+    LLVM_DEBUG(llvm::dbgs() << "Total number of loops: " << result << '\n');
+    return result;
+  }
+};
+
+/// Classical optimization pipeline command-line options. These options are
+/// similar to the ClassicalOptimization pass options, but have different
+/// default settings.
+struct ClassicalOptimizationPipelineOptions
+    : public PassPipelineOptions<ClassicalOptimizationPipelineOptions> {
+  PassOptions::Option<unsigned> threshold{
+      *this, "threshold",
+      llvm::cl::desc("Maximum iterations to unroll. (default: 1024)"),
+      llvm::cl::init(1024)};
+  PassOptions::Option<bool> signalFailure{
+      *this, "signal-failure-if-any-loop-cannot-be-completely-unrolled",
+      llvm::cl::desc(
+          "Signal failure if pass can't unroll all loops. (default: true)"),
+      llvm::cl::init(true)};
```

Comment:
I don't think this option is well-placed. If loops can't be unrolled, they can't be unrolled. That's not an error.

This error condition should be checked in codegen when the assertion "loops are not allowed in this code" is known to be True or False.

Let's not repeat this "feature" in the new pass.